drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Include esp-idf headers first to avoid redefining BIT() macro */
+#include <soc/dport_reg.h>
+#include <soc/i2c_reg.h>
+#include <rom/gpio.h>
+#include <rom/ets_sys.h>
+#include <soc/gpio_sig_map.h>
+
+#include <errno.h>
+#include <gpio.h>
+#include <i2c.h>
+#include <misc/util.h>
+#include <string.h>
+
+#define I2C_ESP32_NUM_CMDS 16
+#define I2C_ESP32_BUFFER_SIZE 32
+#define I2C_ESP32_TIMEOUT_MS 100
+#define I2C_ESP32_SPIN_THRESHOLD 10
+#define I2C_ESP32_YIELD_THRESHOLD (I2C_ESP32_SPIN_THRESHOLD / 2)
+#define I2C_ESP32_TIMEOUT \
+	((I2C_ESP32_YIELD_THRESHOLD) + (I2C_ESP32_SPIN_THRESHOLD))
+
+enum i2c_esp32_opcodes {
+	I2C_ESP32_OP_RSTART,
+	I2C_ESP32_OP_WRITE,
+	I2C_ESP32_OP_READ,
+	I2C_ESP32_OP_STOP,
+	I2C_ESP32_OP_END
+};
+
+struct i2c_esp32_cmd {
+	u32_t num_bytes : 8;
+	u32_t ack_en : 1;
+	u32_t ack_exp : 1;
+	u32_t ack_val : 1;
+	u32_t opcode : 3;
+	u32_t : 17;
+	u32_t done : 1;
+};
+
+struct i2c_esp32_data {
+	union dev_config config;
+	u16_t address;
+
+	struct k_sem fifo_sem;
+	struct k_sem transfer_sem;
+};
+
+struct i2c_esp32_config {
+	const struct {
+		volatile u32_t *ctr;
+		volatile u32_t *fifo_cfg;
```

Comment:
This scheme seems clumsy.  You're building and storing a distinct pointer to every register?  Why not have a "struct esp32_i2c_regs" structure which specifies a single struct with the register array and keep a **single** volatile pointer to that?

Or maybe you're using "pointer to volatile u32_t" as the **type** of the register (which is also wrong: volatile goes on the thing pointed to not the pointer itself)?

Basically: this is a super-giant config struct, and I dont' understand why.